src/axom/primal/geometry/BezierPatch.hpp

```diff
@@ -0,0 +1,875 @@
+            {
+              const int end = ord_u - p;
+              for(int k = 0; k <= end; ++k)
+                dCarray[k] = lerp(dCarray[k], dCarray[k + 1], uv);
+            }
+            c[q][i] = dCarray[0];
+          }
+      }
+      // Run de Casteljau algorithm on each column of control nodes and each dimension
+      else  // Keeping a fixed value of v
+      {
+        c.setOrder(ord_u);
+        axom::Array<T> dCarray(ord_v + 1);  // Temp array
+
+        for(int p = 0; p <= ord_u; ++p)
+          for(int i = 0; i < NDIMS; ++i)
+          {
+            for(int q = 0; q <= ord_v; ++q)
+              dCarray[q] = m_controlPoints(p, q)[i];
+
+            for(int q = 1; q <= ord_v; ++q)
+            {
+              const int end = ord_v - q;
+              for(int k = 0; k <= end; ++k)
+              {
+                dCarray[k] = lerp(dCarray[k], dCarray[k + 1], uv);
+              }
+            }
+            c[p][i] = dCarray[0];
+          }
+      }
+    }
+
+    return c;
+  }
+
+  /*!
+   * \brief Evaluates a Bezier patch at a particular parameter value (\a u, \a v)
+   *
+   * \param [in] u parameter value at which to evaluate on the first axis
+   * \param [in] v parameter value at which to evaluate on the second axis
+   * \return p the value of the Bezier patch at (u, v)
+   *
+   * \note We typically evaluate the patch at \a u and \a v between 0 and 1
+   */
+  PointType evaluate(T u, T v) const
+  {
+    if(getOrder_u() >= getOrder_v())
+      return isocurve(u, 0).evaluate(v);
+    else
+      return isocurve(v, 1).evaluate(u);
+  }
+
+  /*!
+   * \brief Computes a tangent of a Bezier patch at a particular parameter value (\a u, \a v) along \a axis
+   *
+   * \param [in] u parameter value at which to evaluate on the first axis
+   * \param [in] v parameter value at which to evaluate on the second axis
+   * \param [in] axis orientation of vector. 0 for fixed u, 1 for fixed v
+   * \return v a tangent vector of the Bezier patch at (u, v)
+   *
+   * \note We typically find the tangent of the patch at \a u and \a v between 0 and 1
+   */
+  VectorType dt(T u, T v, int axis) const
+  {
+    SLIC_ASSERT((axis == 0) || (axis == 1));
+    if(axis == 0)  // Get isocurve at fixed v
+      return isocurve(v, 1).dt(u);
+    else  // Get isocurve at fixed u
+      return isocurve(u, 0).dt(v);
+  }
+
+  /*!
+   * \brief Computes the normal vector of a Bezier patch at a particular parameter value (\a u, \a v)
+   *
+   * \param [in] u parameter value at which to evaluate on the first axis
+   * \param [in] v parameter value at which to evaluate on the second axis
+   * \return vec the normal vector of the Bezier patch at (u, v)
+   *
+   * \note We typically find the normal of the patch at \a u and \a v between 0 and 1
+   */
+  VectorType normal(T u, T v) const
+  {
+    VectorType tangent_t = dt(u, v, 0);
+    VectorType tangent_s = dt(u, v, 1);
+    return VectorType::cross_product(tangent_t, tangent_s);
```

Comment:
Possible optimization:

```suggestion
    return VectorType::cross_product(dt(u, v, 0), dt(u, v, 1));
```

It might allow the compiler to optimize better, but might not be worth the loss in readability.